### PR TITLE
[@types/mongoose]: strong typed model.create()/new ()

### LINKED ISSUE #41048: @types/mongoose No schema inference on model methods
If you know how to fix the issue, make a pull request instead.

- [x] I tried using the `@types/mongoose` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @horiuchi @lukasz-zak @Alorel @jendrikw @ethanresnick @vologab @jussikinnula @ondratra @alfirin @idandrd @various89 @Fazendaaa @NormanPerrin @danmana @stablio @emmanuelgautier @frontendmonster @mingchen @penumbra1 @orblazer @HughKu @niuware @vladmel1234 @jloveridge @grimmer0125 @richarddd @ToucheSir @borfig @dinodeSimon @anthony-kenikh @iamchathu @tomyam1 @nephix @ChazEpps

I would like to know if there's any plausible reason for not having schema inference on all model methods such as `findOne`, `insert`, etc.

Given the following code:

```js
const userSchema = new Schema({
  name: String,
  email: String,
  password: String,
});

interface UserDocument extends Document {
  name: string;
  email: string;
  password: string;
}

const User = model<UserDocument>('User', userSchema);

// Will never give a compile error even though it's 
// possible to know the fields that should be defined here.
User.create({ emailll: 'some@email.com' });

// Same happens here since the param type is `any`
const user = new User({ dupa: '123' });
```

I would like to have the inference either from the model's schema or from the document interface that we define, but currently all types are `any` when it comes to query conditions or fields to insert, such as `User.create({ notSupposedToBeHere: 'but will be' })` which will either give a run time error due to validation or add the field to the database, but never a type error.

When using nodejs mongodb driver we can easily get the document fields by passing a schema type to the `collection` method, like so `db.collection<{name: string}>('user')` and properly get the `name` field when querying (although it will still allow for any other key to be there), and also get a type error when inserting a field that is not defined, preventing you at compile time to mess your database, so I guess that could be made possible here too.

So, again, is there any reason for this to not be done? 

Thanks!

### PR DESCRIPTION
**Pull request by andreialecu**
Fixes #41048

This PR allows type checking of `Model.create({ ... })` and `new Model()` based on the type of the schema, if one is specified.

If a schema is not specified, it still defaults to `any` so there should be no breaking changes.

Tests were added.

It also allows overriding the type with `.create<CustomType>(doc: CustomType)` for cases where the autodetected type needs to be overridden.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by andreialecu** [comment]
I think if this is merged, it should be published as a new major version of the types. 

It's possible that it breaks projects where types are used incorrectly. 

To clarify, if a schema interface/class/type includes multiple non-optional properties that are not all specified in `.create()`, it will result in an error. See the test for an example of an // $ExpectError when not all required properties are specified.

**Comment by 101arrowz** [comment]
Correct me if I'm wrong, but it appears that this does not account for `Map<K, V>` in documents (which can be specified with `Record<K, V>` or `[K, V][]`). Might need to modify the `DocumentDeclaration` type to add support for them.

File: types/mongoose/test/model.ts
Comment (avaly) [inline types/mongoose/test/model.ts:504]: To workaround the functions inside this type and remove the necessity of the `DeepNonFunctionProperties` wrapper, the following pattern should be used:

```ts
interface FooDocument extends mongoose.Document {
    name: string;
    deeperFuncTest: {
        test: string;
        // ensure this doesn't turn required
        optionalTest?: string;
    };
}
interface FooModel extends mongoose.Model<FooDocument> {
    someFunc: () => any;
    deeperFuncTest: {
        someFunc: () => any;
    };
}

const fooSchema = new mongoose.Schema({
    name: { type: String, required: true },
});

var Foo = mongoose.model<FooModel, FooDocument>("Foo", fooSchema);
```

**Review by avaly — CHANGES_REQUESTED** [review]
Documents should not have function properties. Only models should have such properties.

File: types/mongoose/index.d.ts
Comment (andreialecu) [reply types/mongoose/index.d.ts:100]: Yes, but `FilterQuery` and `UpdateQuery` mark all properties optional, so any functions in the interface is there but not required.

For `Create` it can pop up as required, if it's required in the interface.

I know that in theory this is a code smell when the interface defines a function in it, when used for the schema. 

But... in actual usage it's pretty common to reuse the type of a class as the type of the schema, so this just makes it nice and removes all functions from the type used for creation.

File: types/mongoose/test/model.ts
Comment (andreialecu) [reply types/mongoose/test/model.ts:508]: See my other comment. Let me know what you think.

**Comment by andreialecu** [comment]
@101arrowz thanks for reporting! Just added this, and updated tests.

File: types/mongoose/test/model.ts
Comment (andreialecu) [reply types/mongoose/test/model.ts:511]: For more context on why there could be functions inside models, consider [typegoose](https://github.com/typegoose/typegoose#motivation).

It makes it easy to generate schemas based on decorated classes. Since these classes can potentially also include methods in them, a way to omit them for creation is required.

Otherwise, it leads to additional boilerplate code, where a separate class/interface needs to be defined just to get `.create()` to comply.

**Comment by andreialecu** [comment]
The PR seems to now hopefully be complete. It passes all validation in an actual project I've been working on, including a bunch of edge cases with optional properties, tuples and maps. I've included them in the tests.

**Comment by andreialecu** [comment]
I had to bump minimum typescript version to 3.4 as there was an error on typescript 3.3 with one of the tests due to a bug in that version, I assume:

```
Error: /home/aandrei/projects/DefinitelyTyped/types/mongoose/test/model.ts:527:1
ERROR: 527:1  expect  TypeScript@3.3: Expected an error on this line, but found none.
```
Offending line: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44619/files#diff-fe02473477a238e9eb9f7b0c22653fbaR527

File: types/mongoose/test/model.ts
Comment (andreialecu) [reply types/mongoose/test/model.ts]: Additionally, in practice if you have something like

```ts
class Foo {
  bar: Map<string, Bar>;
  privateProp: string;

  setBar(barLike: any) {
    this.bar = convertSomehow(someBar);
  }
}

class Bar {
  title: string;
  
  someFunction(someParam: any) {
    
  }
}
```

Now you need interfaces that exclude functions for both `Foo` and `Bar` and you need to connect `FooSchema.bar` to the `BarSchema` type. 

The more complex and intertwined the models, the harder this becomes to do, up to a point where it becomes unmaintainable.

I feel the right thing to do is just omit functions by default in the type. When using plain mongoose they don't matter anyway.

Comment (avaly) [reply types/mongoose/test/model.ts]: I'm not sure I am comfortable with this change, so I'll leave other reviewers take a stab at this.

Comment (andreialecu) [reply types/mongoose/test/model.ts]: FWIW, it's not really a "change", but an addition. :) Nothing about current behavior really changes.

**Comment by jeremyben** [comment]
I tried this once, but you will get problems with getters and setters:
```ts
interface DocWithSettersGetters extends mongoose.Document {
  readonly getterOnly: string
  getterAndSetter: string
}
```
This is how you declare getters and setters in interfaces.
It's not possible to detect and remove those properties.

**Comment by andreialecu** [comment]
@jeremyben thanks for the review. Those cases should be an exception however, and I implemented an escape hatch:

You can do this:
```ts
model.create<Omit<DocWithSettersGetters, "getterOnly" | "getterAndSetter">>(...)
```
or
```ts
model.create<Partial<DocWithSettersGetters>>()
```

File: types/mongoose/test/model.ts
Comment (avaly) [reply types/mongoose/test/model.ts]: I know :)

Could you perhaps use these utility types from `mongodb` instead of your `DeepNonFunctionProperties` utility type?

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mongodb/index.d.ts#L1239-L1246

**Comment by andreialecu** [comment]
@jeremyben It should be possible to remove the readonly properties as well, but for a first implementation I think this is good enough.

Here's a https://github.com/krzkaczor/ts-essentials#writablekeys type that could be used to omit non-writable keys.

For this PR, take a quick look at the tests for all supported scenarios:

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44619/files#diff-fe02473477a238e9eb9f7b0c22653fba

It seems like methods could be very common, and those are automatically removed. I think readonly properties and getters/setters would be relatively rare.

**Comment by andreialecu** [comment]
I've prepared a .diff patch for [patch-package](https://www.npmjs.com/package/patch-package):
https://gist.github.com/andreialecu/fb9b58a15705729eb107a706abcdd6d2

Should make it really easy to apply and test on a real project. Seems to work fine on the ones I've been working on.

**Comment by 101arrowz** [comment]
In [the PR that inspired this one](https://github.com/typegoose/typegoose/pull/241), the solution for getters was a `ReadOnlyKeys<T>` type, consider using that to automatically exclude getters. For getter AND setter properties, there is no solution but to omit.

Also, in my personal opinion, making all keys optional defeats the purpose of this PR.

**Comment by andreialecu** [comment]
@jeremyben @101arrowz The latest commit automatically removes all readonly properties from the type used for creation. Should address both issues.

Take a look at the tests to see what's supported. 

Unfortunately, one of the constructs only works with TS3.7+ so functions and read only properties will not be properly removed from `TValue` in `Map<TKey, TValue>`, but they are removed if supplied as plain objects. (see tests)

Credits to @101arrowz for taking this on in the typegoose repository which made me think of it again. 

I have tried adding strong types for `.create()` to `@types/mongoose` before but I couldn't find a way to keep backwards compatibility until recently, through the logic here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44619/files#diff-f77e4c92e873042528b4ef79285cdf3cR185-R190

**Comment by 101arrowz** [comment]
Any updates on this? Will this PR be considered for merging?

**Comment by amcasey** [comment]
> I think if this is merged, it should be published as a new major version of the types.

The major and minor versions have to stay in sync with the underlying library - they do not convey semver info about the types.

> It's possible that it breaks projects where types are used incorrectly.

Generally, bug fixes are considered to be acceptable breaks.  Having said that, this is a popular library, so it would be nice to minimize friction for existing users.

**Comment by amcasey** [comment]
My concern with this PR is that the types are very complicated (and expensive-looking - I'm surprised the bot didn't complain) and there still seem to be unresolved concerns from earlier reviewers.  If this were a less popular package, I'd say "go for it", but it's very easy to create difficult-to-trace editor experience regressions with changes like this one.

So what now?  I'd really like to get some kind of confirmation from a mongoose domain expert that the change makes sense and, ideally, a report from a mongoose user who actually worked against these types for a bit.

**Comment by andreialecu** [comment]
@amcasey the breaking changes are minimal and will only affect projects where types are used wrong in the first place. As an example, this would introduce minor breaks when a schema property is marked required in the interface, and is then not supplied in `.create()` calls.

The fix for that is to mark the property optional in the interface, as it should've been in the first place, by adding a `?`.

If a specific schema type is not supplied, then the PR doesn't affect anything, the entire schema is treated as `any`.

I'm not sure what the unresolved complaints you refer to are. Everything is addressed, only the complexity is (subjectively) high, but that's a requirement in order to support removing unnecessary properties, so it can work on par with vanilla mongoose.

I've been using this in production for the past few weeks with no problems. And I know others who have too.

**Comment by jussikinnula** [comment]
I looked the changes, and could we actually make the type parameter to default to "any", so it wouldn't actually break even minimally?

I'm anyway happy with or without the default.

**Comment by andreialecu** [comment]
@jussikinnula the type parameter does default to `any`, feel free to try it in an actual real-life project.

You can easily use [patch-package](https://www.npmjs.com/package/patch-package) to apply this as a patch.

If you find any breakage introduced simply by swapping them, let me know. The definitions evolved over multiple iterations, and there are lots of edge cases covered so I don't think anyone should run into any issues. 

Even if they do, you can opt-out via `.create<TOther>({ })`, where you can use `any` if you want, `Omit<>` various properties, or define a different type entirely.

**Comment by jeremyben** [comment]
I'm in favor of a more gradual improvement to the create signature.
We're going from `any` to strict generic constraint with a lot of conditional and recursive inference.

Like I already suggested, I think a PR with less recursivity/complexity and making it `Partial` would be better and less prone to breaking.
As a first baby step towards more strictness. And then iterate on it.

Strictness is what we want ideally, but even mongoose itself at runtime does not offer very strict abstractions, unfortunately.

**Comment by andreialecu** [comment]
I personally disagree strongly about making it `Partial<>` since that wouldn't be idiomatic typescript, and just breaks the contract between the schema and the actual run-time behavior of mongoose. Basically leading to further efforts and lost time while debugging things.

I feel that, considering how important and popular these types are, we are responsible for enforcing good practices. 

There is nothing good about having required properties in the schema interface if that are not actually required. The proper thing would be to mark them optional - which is the only minor "breaking" usage introduced by the PR. 

The recursivity and complexity of the type is just an implementation detail. All the complexity is just for edge cases that affect a minor percentage of schemas, but once you do get affected, I think it's nice for things to "just work". 

For 99% of cases, the complexity is not important, as those recursive conditionals will never be activated. 

The bot confirms that at worst, the performance is exactly the same.

And again, as an escape hatch, `.create<Partial<Model>>(...)` would work to make everything `Partial`.

**Comment by jeremyben** [comment]
What I suggested was a first step, not the end of it. Simply see how many complains there is, and move foreward with more strictness.

You're right about contract and everything, but Mongoose is itself so loose that Typescript can't simply deal with all that looseness, look at the other signatures. Contract between the type system and runtime behavior is broken even in the find queries (does `.select()` actualy `Pick` or `Omit` properly ? does `.populate()` change the resulting interface accordingly ? I could go on).

This PR is open for 20 days, and worries that it might break existing users, which is the most important thing there. I'm actually trying to soothe those worries, to make this kind of PR happen because I'm also a `strict: true` guy. Like @amcasey, I'm waiting for more feedback, and since I have inherited a complex (and poorly written) mongoose codebase, I'm willing to try it.

About type recursivity, ok I agree with you.

**Comment by andreialecu** [comment]
I agree that in general Mongoose is loose and it's hard to cover all edge cases with proper types.

But `.create()` is actually one of those lucky cases where prior and future state is not important, and where the proper type used for creation can be inferred almost 100% of the time. 

`Create` doesn't induce any side effects, in contrast with the other methods you mentioned, and is probably the only method where non-optional properties make sense.

**Comment by jussikinnula** [comment]
> @jussikinnula the type parameter does default to `any` if no schema type is defined, feel free to try it in an actual real-life project.

Yes, I think I should do that. I have to admit, that the last time I used Mongoose typings in real-life project I had to do always cast the response explicitly. So this is really an improvement. And I tested in simple type that yes it's like you said that any is the default.

I was just thinking that `any[]` (e.g. array of any) would be better for the type that had `any[]` before the patch, but `CreateQuery<TCreate>[]` actually makes it already an array type. So it's good as is.

**Comment by amcasey** [comment]
Still no sign-offs but I'm hearing a tentatively positive sentiment?

**Comment by orta** [comment]
Yeah, my ambiguity on those perf results feel like we should have a test on a real world project - can someone please give this a run? ( you can manually edit the `@types/mongoose` module in your repo to the version in this PR to test)

**Comment by andreialecu** [comment]
@orta is there a benchmark I can run on a real world project? Can I run the DT benchmark just on my project, to compare type performance somehow?

As mentioned before I'm using these types, and a few other people I've talked to on Discord on mongoose/typegoose related channels have been trying and using them too without issues. They don't know about the latest discussion here though to report.

Again, `.create()` is only a very minor call-site for mongoose apps. I don't think there would be more than a handful such calls even in complex projects. And, most schemas are not complex enough for most of those recursive conditionals to ever be entered. However, I'm not aware of the internals of the typescript compiler to know if they could have an additional performance related side-effect just by being there, and not actually executing to their full extent.

**Comment by amcasey** [comment]
@andreialecu If you have a code base that consumes the `.create` method, it should be sufficient to build it using tsc and `--extendedDiagnostics`.  If the check time doesn't substantially increase (compared to consuming the types before your change), it's probably fine.

**Comment by andreialecu** [comment]
@amcasey:

Pre patch:
```
rm -rf ./dist && yarn tsc --extendedDiagnostics
yarn run v1.22.0
$ /home/aandrei/projects/projectsport-server/node_modules/.bin/tsc --extendedDiagnostics
Files:                        1362
Lines:                      539256
Nodes:                     1764606
Identifiers:                651735
Symbols:                    580361
Types:                      230563
Memory used:               877475K
Assignability cache size:   312203
Identity cache size:           506
Subtype cache size:           1110
Strict subtype cache size:     378
I/O Read time:               0.10s
Parse time:                  2.26s
Program time:                3.29s
Bind time:                   1.14s
Check time:                  7.88s
transformTime time:          0.58s
commentTime time:            0.00s
printTime time:              0.98s
Emit time:                   1.00s
Source Map time:             0.03s
I/O Write time:              0.02s
Total time:                 13.31s
Done in 14.12s.
```

Post patch:
```
rm -rf ./dist && yarn tsc --extendedDiagnostics
yarn run v1.22.0
$ /home/aandrei/projects/projectsport-server/node_modules/.bin/tsc --extendedDiagnostics
Files:                        1362
Lines:                      539324
Nodes:                     1765122
Identifiers:                651931
Symbols:                    587530
Types:                      249509
Memory used:               888572K
Assignability cache size:   317176
Identity cache size:           879
Subtype cache size:           1144
Strict subtype cache size:     430
I/O Read time:               0.11s
Parse time:                  2.45s
Program time:                3.48s
Bind time:                   1.15s
Check time:                  8.05s
transformTime time:          0.61s
commentTime time:            0.00s
printTime time:              1.01s
Emit time:                   1.02s
Source Map time:             0.01s
I/O Write time:              0.02s
Total time:                 13.70s
Done in 14.48s.
```

**Review by amcasey — APPROVED** [review]
In the absence of concrete objections or evidence of a performance regression, I'm going to go ahead and merge this.

**Comment by bopfer** [comment]
@andreialecu If you have some free time, can you check out this issue? https://github.com/DefinitelyTyped/DefinitelyTyped/issues/45305

I am a little stuck at the moment on what is required to change in my code base as a result of this PR that adds type checking for `create`.

**Comment by andreialecu** [comment]
hey @bopfer, it's possible you're hit by https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45355 which is waiting for approval

There was an issue with arrays with the changes in this PR.

**Pull request merged**
